File: writeups/206_Reverse_Linked_List/ruby/main.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
  return [] if head.nil?

  current_node = head
  stack_nodes = [nil]
  while current_node.next do
    stack_nodes.push(current_node)
    current_node = current_node.next
  end
  head_node = current_node
  while !stack_nodes.empty?
    current_node.next = stack_nodes.pop
    current_node = current_node.next
  end

  head_node
end

File: writeups/206_Reverse_Linked_List/ruby/typescript/main.ts
function reverseList(head: ListNode | null): ListNode | null {
  if (head === null) return null;

  let currentNode: ListNode = head;
  const stackNodes: (ListNode | null)[] = [null];
  while (currentNode.next) {
    stackNodes.push(currentNode);
    currentNode = currentNode.next;
  }
  const headNode: ListNode = currentNode;
  while (stackNodes.length) {
    currentNode.next = stackNodes.pop();
    currentNode = currentNode.next;
  }

  return headNode;
}
